### ISSUE TITLE
Lazy Loading with Router while using libraries in different modules

### ISSUE THREAD

**Issue by paulheising** [angular/angular#45304]
### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

No

### Description

Actually I'm not sure, if it's a bug, or a known drawback and I'm actually having false expectations. Just let me explain.

- We have an angular workspace that contains:
  - `foo-lib`: A library that is going to be used in `foo-app` (below)
    - Contains `AlphaModule`, which declares and exports `AlphaComponent`
    - Also contains `BetaModule`, which declares and exports `BetaComponent`
    - Both being exported via the lib's public-api.ts
  - `foo-app`: An Application that uses the router for lazy loading
    - theres an `AppModule`, an eagerly loaded startup module of the app
      - The `AppComponent`, declared and bootstrapped by the `AppModule`, only uses the `AlphaComponent` from the lib above
    - and theres a `LazyModule`, which is loaded lazily in the `AppRoutingModule` as described here: https://angular.io/guide/lazy-loading-ngmodules
      - The `LazyComponent`, declared and routed to via `LazyRoutingModule`, only uses the `BetaComponent` from the lib above

## Expectation:

Following the description above, when I build the app, I expect the final JS bundles...
- to have the `AlphaComponent` only referenced in the main.js bundle
- and to have the `BetaComponoent` only referenced in the lazy.js bundle

Because that's the point of doing lazy loading huh? Having stuff, that is not required in the main bundle, moved to their respective lazy loaded bundles, where it's actually needed

## Actual behaviour

In our production app, we once noticed a huge amount of unused JS inside our main bundle, reported by core web vitals measurements tool.

Further analysis identified an internal angular component library of ours as main problem and revealed that the expectation written above is actually false:

- As long as we use the lib *only* in our lazy module, everything is fine. The stuff we reference from the lib only appears in the lazy.js bundle
- But as soon as we use stuff from the lib also in our AppModule or components that directly or indirectly belong to it, everything else from the lib which is only used in lazy modules, also goes to the main.js bundle when compiling
- Referencing the example drawn above that means
  - The main.js bundle actually contains both `AlphaComponent` and `BetaComponent` from the lib, although `BetaComponent` is not used in any way in the `AppModule`, but only in the `LazyModule`
  - When we remove all `AlphaComponent` and `AlphaModule` references from our `AppModule` and components that belong eagerly to it, we find out that the `BetaComponent` now is actually contained in the lazy bundle, and not in the main bundle any more, that part follows my expectation.

## Workaround

However, the things written above are only true as long as the library only has one single entry point. This is the case for our internal component library. Each individual component is wrapped inside its own NgModule and most of them have no cross references between each other, it's mainly a collection of reusable ui-components. So we can always import only those NgModules we actually need, always thought we'd rely on tree shaking to reduce the bundle to what's actually being used (which is true unless you use lazy loading and reference the lib from different routes)

As soon as we create seperate secondary entry points for each entity in the lib which is meant to be used independently, the problem does not occur any longer.

I also tried that with non-angular libs (normal npm package with some typescript stuff being exported) - same issue.

## What is this?

Is it a bug?

Is it a problem, may it be known or unknown, which exists but is currently not resolvable (unless using sepearte entry points in the lib)? If that's the case, I think it should be mentioned in the docs about lazy loading, otherwise I don't think most devs are aware of this problem and just run into the same trap as we did. Blindly and unconciously.

### Please provide a link to a minimal reproduction of the bug

https://github.com/paulheising/demo-lazy-loading-with-libs

### Please provide the environment you discovered this bug in (run `ng version`)

```true
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 13.2.5
Node: 14.19.0
Package Manager: npm 6.14.16
OS: linux x64

Angular: 13.2.6
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1302.5
@angular-devkit/build-angular   13.2.5
@angular-devkit/core            13.2.5
@angular-devkit/schematics      13.2.5
@angular/cli                    13.2.5
@schematics/angular             13.2.5
ng-packagr                      13.2.1
rxjs                            7.5.5
typescript                      4.5.5
```

**Comment by JoostK** [comment]
> Is it a problem, may it be known or unknown, which exists but is currently not resolvable (unless using sepearte entry points in the lib)?

Indeed, this is a limitation of Webpack which does not do "code motion", i.e. it does not split ES modules—and entry-points are compiled into a single FESM (Flat ES Module)—according to how symbols are used to improve code-splitting granularity. As you mention, this is where secondary entry-points avoid this limitation.

**Comment by alxhub** [comment]
This is documented in the [Angular Package Format](https://angular.io/guide/angular-package-format#entrypoints-and-code-splitting) specification, actually. You're exactly right that one of the main reasons libraries may wish to expose secondary entry points is to provide flexibility for code-splitting.

This is a tradeoff that we've had to make. Bundlers like Webpack implement ES modules via function wrappers and other techniques that add code size overhead for each module. This overhead is significant enough that the benefits of flattening far outweigh the constraints it creates around code splitting.

The JavaScript build/bundling ecosystem is constantly changing, and there is a potential for this tradeoff to change over time. For this reason, the APF in v13 actually ships unflattened builds as well. They're unused by the CLI currently, but we included them to enable future experiments around different ways of bundling Angular code.

**Comment by paulheising** [comment]
Thanks for clearing this up!

Then we'll go and split our lib up into seperate entry points. One of those tasks I wish we'd have done from the start - because now it's going to be a lot of donkeywork :smile:

**Comment by alan-agius4** [comment]
Closing as this is already documented as per @alxhub comment.
